src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt

```diff
@@ -0,0 +1,235 @@
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.ui.RefactoringDialog
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import com.intellij.ui.components.JBTextField
+import com.intellij.ui.layout.panel
+import com.intellij.usageView.BaseUsageViewDescriptor
+import com.intellij.usageView.UsageInfo
+import com.intellij.usageView.UsageViewDescriptor
+import com.intellij.util.ui.JBUI
+import org.jetbrains.annotations.TestOnly
+import org.rust.ide.utils.GenericConstraints
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.openapiext.addTextChangeListener
+import org.rust.openapiext.isUnitTestMode
+import javax.swing.JComponent
+import javax.swing.JTextField
+
+class RsExtractTraitHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        val offset = editor.caretModel.offset
+        val element = file.findElementAt(offset) ?: return
+        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
+        if (impl.traitRef != null) return
+        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+
+        val members = (impl.members ?: return).childrenOfType<RsItemElement>()
+        if (members.isEmpty()) return
+        val memberInfos = members.map { RsMemberInfo(it, false) }
+
+        val dialog = RsExtractTraitDialog(project, impl, memberInfos)
+        if (isUnitTestMode) {
+            dialog.doAction()
+        } else {
+            dialog.show()
+        }
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        /* not called from the editor */
+    }
+}
+
+class RsExtractTraitDialog(
+    project: Project,
+    private val impl: RsImplItem,
+    private val memberInfos: List<RsMemberInfo>,
+) : RefactoringDialog(project, false) {
+
+    private var traitNameField: JTextField = JBTextField().apply {
+        addTextChangeListener { validateButtons() }
+    }
+
+    init {
+        super.init()
+        title = "Extract Trait"
+        validateButtons()
+    }
+
+    override fun createCenterPanel(): JComponent = panel {
+        blockRow {
+            cell(isFullWidth = true) {
+                label("Trait name:")
+            }
+            traitNameField().focused()
+        }
+        row {
+            val members = RsMemberSelectionPanel("Members to form trait", memberInfos)
+            members.minimumSize = JBUI.size(0, 200)
+            members.table.addMemberInfoChangeListener { validateButtons() }
+            members()
+        }
+    }
+
+    override fun validateButtons() {
+        super.validateButtons()
+        previewAction.isEnabled = false
+    }
+
+    override fun areButtonsValid(): Boolean =
+        isValidRustVariableIdentifier(traitNameField.text) && memberInfos.any { it.isChecked }
+
+    @VisibleForTesting
```

Comment:
I'm not sure if this annotation makes sense here. Looks like this method is public and also used deep inside the platform (outside the tests), WDYT?